ql/src/java/org/apache/hadoop/hive/ql/Compiler.java

```diff
@@ -329,7 +357,34 @@ private void recordValidTxns(HiveTxnManager txnMgr) throws LockException {
     LOG.debug("Encoding valid txns info " + txnStr + " txnid:" + txnMgr.getCurrentTxnId());
   }
 
-  private QueryPlan createPlan(BaseSemanticAnalyzer sem) {
+  protected QueryPlan createPlan(BaseSemanticAnalyzer sem) {
+    if (getStatementType(tree) == StatementType.EXECUTE) {
+      assert (tree != null);
+      String queryName = PreparePlanUtils.getPrepareStatementName(tree);
+
+      SessionState ss = SessionState.get();
+      if (ss != null) {
+        if (ss.getPreparePlans().containsKey(queryName)) {
+          QueryPlan plan = ss.getPreparePlans().get(queryName);
+          QueryPlan planCopy = PreparePlanUtils.makeCopy(plan);
+          bindDynamicParams(planCopy, getParameterMap());
+          // This is no longer prepare plan
+          planCopy.setIsPrepareQuery(false);
+          //TODO: this will create empty schema, need to figure out a way to create proper schema
```

Comment:
Once we have retrieved the plan at this stage, we probably need to change query id to avoid any issue, i.e., multiple queries with same query id? Also we could update query text in plan and I am not sure if anything else is needed.